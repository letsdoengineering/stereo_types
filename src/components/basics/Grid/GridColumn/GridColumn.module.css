.grid-column {
    display: block;
    flex: none;
    flex-direction: column;
}

.grid-column--padded {
    padding-left: 0.6rem;
    padding-right: 0.6rem;
}

.size-auto {
    flex-basis: 0;
    flex-grow: 1;
    flex-shrink: 1;
}

.size-1 {
    width: calc(100% / 12);
}

.size-2 {
    width: calc(100% / 6);
}

.size-3 {
    width: calc(100% / 4);
}

.size-4 {
    width: calc(100% / 3);
}

.size-5 {
    width: calc(100% / 2.4);
}

.size-6 {
    width: calc(100% / 2);
}

.size-7 {
    width: calc((100% / 12) * 7);
}

.size-8 {
    width: calc(100% / 1.5);
}

.size-9 {
    width: calc((100% / 12) * 9);
}

.size-10 {
    width: calc(100% / 1.2);
}

.size-11 {
    width: calc((100% / 12) * 11);
}

.size-12 {
    width: 100%;
}
