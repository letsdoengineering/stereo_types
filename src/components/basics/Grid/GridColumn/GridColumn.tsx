import React from 'react'
import getClassNames from '../../../../utils/get-class-names'
import * as gridColumnStyles from './GridColumn.module.css'

type GridColumnProps = React.HTMLAttributes<HTMLDivElement> & {
  /** optional: specify how much space in a row a column occupies in increments of 1/12ths */
  size?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12'
  /** optional: class-names */
  className?: string
  /** indicates whether or not the row container should be padded */
  isPadded?: boolean
  /** children always of type ReactNode */
  children?: React.ReactNode
}

/** GridColumn: component that returns a div styled as a grid column with
 * automatic sizing or sizing specified by size prop*/
const GridColumn: React.FC<GridColumnProps> = ({
  size,
  isPadded = false,
  className,
  children,
  ...rest
}: GridColumnProps) => {
  const classNames = getClassNames({
    defaultClasses: [
      `${gridColumnStyles.gridColumn}`,
      size ? `${gridColumnStyles[`size${size}`]}` : `${gridColumnStyles.sizeAuto}`,
    ],
    className,
    conditionalClasses: {
      [`${gridColumnStyles.gridColumnPadded}`]: isPadded,
    },
  })

  return (
    <div className={classNames} {...rest}>
      {children}
    </div>
  )
}

export default GridColumn
