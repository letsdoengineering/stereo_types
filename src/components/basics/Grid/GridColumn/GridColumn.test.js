import * as React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import GridColumn from './GridColumn'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'

describe('GridColumn', () => {
    describe('dynamic classes', () => {
        it('renders a grid column with sizeAuto class when size not specified', async () => {
            render(<GridColumn className={'my-random-class'}>Some column</GridColumn>)
            const gridColumn = await screen.getByText('Some column')

            await waitFor(() => {
                expect(gridColumn).toHaveClass('gridColumn sizeAuto my-random-class')
                expect(gridColumn).not.toHaveClass('size1 size2 size3')
            })
        })

        it("renders a grid column with an 'is-n' class where n is only the integer specified in the size prop", async () => {
            render(
                <GridColumn className={'my-random-class'} size='1'>
                    Some column
                </GridColumn>
            )
            const gridColumn = await screen.getByText('Some column')

            await waitFor(() => {
                expect(gridColumn).toHaveClass('gridColumn size1 my-random-class')
                expect(gridColumn).not.toHaveClass('size2 size3 size4')
            })
        })
    })

    it('has not changed the JSX layout of the page from previous snapshot', () => {
        const tree = renderer.create(<GridColumn>Some column</GridColumn>).toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(<GridColumn>Some column</GridColumn>)
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
