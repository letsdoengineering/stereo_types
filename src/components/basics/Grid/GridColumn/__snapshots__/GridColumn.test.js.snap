// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridColumn has not changed the JSX layout of the page from previous snapshot 1`] = `
<div
  className="gridColumn sizeAuto"
>
  Some column
</div>
`;
